### PR TITLE
fix(backend/external-api): Improve security & reliability of API key storage

### PR DESCRIPTION
**Pull request by Pwuts**
Our API key generation, storage, and verification system has a couple of issues that need to be ironed out before full-scale deployment.

### Changes 🏗️

- Move from unsalted SHA256 to salted Scrypt hashing for API keys
- Avoid false-negative API key validation due to prefix collision
- Refactor API key management code for clarity
- [refactor(backend): Clean up API key DB & API code (#10797)](https://github.com/Significant-Gravitas/AutoGPT/pull/10797)
  - Rename models and properties in `backend.data.api_key` for clarity
  - Eliminate redundant/custom/boilerplate error handling/wrapping in API
key endpoint call stack
  - Remove redundant/inaccurate `response_model` declarations from API key
endpoints

Dependencies for `autogpt_libs`:
- Add `cryptography` as a dependency
- Add `pyright` as a dev dependency

#### For code changes:
- [x] I have clearly listed my changes in the PR description
- [x] I have made a test plan
- [x] I have tested my changes according to the test plan:
  - Performing these actions through the UI (still) works:
    - [x] Creating an API key
    - [x] Listing owned API keys
    - [x] Deleting an owned API key
  - [x] Newly created API key can be used in Swagger UI
  - [x] Existing API key can be used in Swagger UI
  - [x] Existing API key is re-encrypted with salt on use

### CODE REVIEW

**Comment by AutoGPT-Agent** [comment]
Thank you for this security improvement PR! Moving from unsalted SHA256 to salted Scrypt is a solid enhancement for our API key system.

### What looks good
- The implementation is well-structured with proper testing
- The migration path for legacy keys is handled elegantly
- Code is well-documented with clear function purposes
- The Prisma schema changes align well with the code changes

### What needs attention
- **Missing test plan**: Since this is a security-related change, we need a clear test plan. Please complete the test plan section in the PR description with steps to validate the changes. For example:
  - Testing API key creation with the new system
  - Verifying that legacy keys still work
  - Confirming migration from legacy to new format works correctly
  - Testing invalid keys are properly rejected

- Consider adding some documentation about the security improvements for future reference

Once you've added a complete test plan, this PR should be ready for approval.

**Comment by AutoGPT-Agent** [comment]
Thanks for this important security improvement to the API key system!

The PR looks well-structured and implements a more secure approach by:
- Moving from unsalted SHA256 to salted Scrypt hashing (much more resistant to brute force attacks)
- Adding safeguards against false-negative API key validation due to prefix collision
- Including a migration path for existing API keys
- Adding a comprehensive test suite for the new functionality

The code quality is good, with clear naming and proper error handling. I appreciate the thorough refactoring for clarity.

Before we can merge this:
1. Please complete the test plan section of the PR checklist. For a security-related change like this, we need a clear plan for how you've verified the changes work correctly, especially the migration path for existing API keys.

2. Consider adding some information about backward compatibility - will existing API keys continue to work? Your code suggests yes (with the legacy key verification), but it would be good to make this explicit in the PR description.

Otherwise, this is a valuable improvement to our API key security model!

**Comment by AutoGPT-Agent** [comment]
Thank you for your PR to improve API key security! Moving from unsalted SHA256 to salted Scrypt hashing is an excellent improvement for our key storage system.

### What looks good
- The code changes are well-structured and include a clear migration path for existing keys
- You've added comprehensive unit tests for the new functionality
- The changes to the database schema are properly handled with a migration script
- Your implementation properly handles user_id validation in all relevant functions
- The refactoring improves code clarity and security

### Issues to address

**Before merging, please complete the checklist:**
- You need to create and execute a test plan for these changes
- Security-related changes especially need verification in a test environment

Please outline your testing approach and confirm you've tested:
1. Generation of new API keys with the improved security
2. Validation of these new secure keys
3. Verification that existing keys continue to work (migration path)
4. Any performance impacts of the more intensive hashing algorithm

Once you've completed the testing, please update the PR description with your test plan and results.

**Comment by AutoGPT-Agent** [comment]
Thank you for this security improvement to our API key system! The changes look solid - moving from unsalted SHA256 to salted Scrypt hashing is a significant security upgrade.

However, I can't approve this PR yet because the test plan is missing. For security-related changes especially, we need a clear test plan that verifies:

1. New API keys are properly generated with the secure hashing
2. Existing API keys still work (backward compatibility)
3. The migration process for legacy keys works correctly
4. API key verification functions properly in all cases

You've added unit tests which is great, but please fill out the test plan checklist with steps to manually verify that the changes work as expected in the real application.

Some specific items to consider for your test plan:
- Generate new API keys through the interface and verify they work
- Test with existing API keys to ensure they still function
- Verify that revocation and suspension still work correctly
- Check that permissions are properly enforced

Once you've completed and checked off the test plan, I'll be happy to approve this PR.

**Comment by AutoGPT-Agent** [comment]
Thank you for this security-focused PR! The improvements to API key storage using salted Scrypt hashing are a significant upgrade over the current unsalted SHA256 approach.

### What looks good:
- The code changes are well-structured and include a migration path for existing keys
- You've added comprehensive tests for the new implementation
- The changes maintain existing user_id security checks
- The database migration script is included

### What needs attention:
- **Test plan**: The PR checklist is incomplete - you've checked that you've listed changes, but you haven't provided a test plan. For security-related changes, it's important to document how you've verified that:
  - Existing API keys continue to work (backward compatibility)
  - New API keys are properly secured
  - The migration from old to new format works reliably

Could you please add a test plan describing how you've tested these changes? Even though you've added unit tests, it would be helpful to know what manual verification you've done with the implementation.

Overall, this looks like a solid improvement to our security posture!

File: autogpt_platform/autogpt_libs/autogpt_libs/api_key/keysmith.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        """
+        Verify an API key against a known hash (+ salt).
+        Supports verifying both legacy SHA256 and secure Scrypt hashes.
+        """
+        if not provided_key.startswith(self.PREFIX):
+            return False
+
+        # Handle legacy SHA256 hashes (migration support)
+        if known_salt is None:
+            legacy_hash = hashlib.sha256(provided_key.encode()).hexdigest()
```
[the remark this answers is not in the document]
Comment (Pwuts) [reply autogpt_platform/autogpt_libs/autogpt_libs/api_key/keysmith.py:48]: It's being phased out as of this PR

**Comment by AutoGPT-Agent** [comment]
Thank you for this security improvement PR! Moving from unsalted SHA256 to salted Scrypt hashing is definitely a good security upgrade, and I appreciate the attention to detail in handling legacy key migration.

### Required Changes

- **Test Plan**: For security-related changes like this, a test plan is essential. Please complete the test plan section in your PR description that outlines how you've tested these changes, especially:
  - Creating new API keys with the new security system
  - Verifying existing keys continue to work (migration path)
  - Testing with invalid keys to ensure proper rejection

### Observations

- You've done a great job with the implementation by adding support for legacy keys, which provides a smooth migration path.
- The test coverage for the new `keysmith.py` looks comprehensive.
- The database migration script is well-structured for the column renames and additions.
- The dependency update for cryptography is properly reflected in both autogpt_libs and backend.

Once you complete the test plan section, this PR should be ready for merge. The security improvements are valuable and the implementation approach is solid.

**Comment by AutoGPT-Agent** [comment]
Thanks for this security improvement PR! The changes look solid and the code improvements make sense.

### Strong points
- Good implementation of salted Scrypt hashing instead of unsalted SHA256
- Comprehensive test coverage for the new implementation
- Migration support for legacy keys
- Well-structured code refactoring with clearer naming
- Database migration script included

### Items to address
- The PR checklist is incomplete. Since this is a security-related change, please complete the test plan section, detailing how you've verified that:
  - New API keys are generated securely
  - Legacy keys still work
  - Invalid keys are properly rejected
  - The migration path for existing keys works correctly

Once you've filled out the test plan and verified these security-critical aspects, this PR should be ready to merge. The code changes themselves look well-implemented and represent a significant security improvement.

**Comment by AutoGPT-Agent** [comment]
Thank you for the PR improving the API key security! This is a valuable enhancement to our security posture, moving from unsalted SHA256 to the more secure salted Scrypt hashing approach.

### What looks good
- Strong implementation of cryptographically secure hashing with Scrypt
- Adding proper salt handling for key storage
- Good test coverage for the new keysmith functionality
- Smart handling of legacy key format migration
- Thorough database migration script for the schema changes

### Before we can merge

**Main blocker:** You haven't completed the test plan section of the PR checklist. For a security-related change like this, we need to understand:

1. How you've verified the changes work correctly
2. What kind of testing you've done to ensure the migration path works for existing API keys
3. Any manual validation steps you've taken

Please add a test plan to the PR description and check off the related items in the checklist. This should outline how you've tested both the new key generation and the migration of existing keys.

**Additional suggestions:**
- Consider adding a comment in the `_migrate_key_to_secure_hash` function explaining when/how this will be triggered
- It might be helpful to document the Scrypt parameters you chose (n=2^14, r=8, p=1) and why these are appropriate security/performance tradeoffs for our use case

Once you've added the test plan, this PR should be ready for a final review and merge.

**Review by ntindle — APPROVED** [review]

**Comment by AutoGPT-Agent** [comment]
Thanks for this security improvement PR! It looks very comprehensive and addresses the important security concerns with API key storage.  

I like the changes you've made:
- Moving to salted Scrypt hashing is a significant security improvement
- The automatic migration of legacy keys on use is a good approach
- The code refactoring makes the API key management clearer and more maintainable
- The tests are thorough and cover both new keys and migration scenarios

However, I noticed your checklist shows that testing with Swagger UI is still incomplete. Before merging, please ensure:

1. Complete the remaining test cases in your checklist, particularly verifying that both new and existing API keys work correctly with Swagger UI
2. Double-check that the API key migration logic is robust against edge cases

Also, since this is a security-related change, it might be worth documenting somewhere (perhaps in a comment) what the security improvement is and why it matters.

Overall, this is a great improvement to the system's security posture. Once the testing is complete, this looks ready to merge.

**Review by Bentlybro — APPROVED** [review]
